src/config/cors.rs

```diff
@@ -0,0 +1,357 @@
+use anyhow::{Error, Result};
+use serde::Deserialize;
+use std::convert::TryFrom;
+
+/// CORS (Cross Origin Resource Sharing) configuration for the HTTP/S
+/// server.
+///
+/// `CorsConfig` holds the configuration for the CORS headers for a
+/// HTTP/S server instance. The following headers are supported:
+///
+/// Access-Control-Allow-Credentials header
+/// Access-Control-Allow-Headers header
+/// Access-Control-Allow-Methods header
+/// Access-Control-Expose-Headers header
+/// Access-Control-Max-Age header
+/// Access-Control-Request-Headers header
+/// Access-Control-Request-Method header
+///
+/// Refer to CORS here: https://www.w3.org/wiki/CORS
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct CorsConfig {
+    /// The Access-Control-Allow-Credentials response header tells browsers
+    /// whether to expose the response to frontend JavaScript code when the
+    /// request's credentials mode (Request.credentials) is include.
+    ///
+    /// The only valid value for this header is true (case-sensitive). If you
+    /// don't need credentials, omit this header entirely (rather than setting
+    /// its value to false).
+    ///
+    /// Source: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials
+    pub(crate) allow_credentials: bool,
+    /// The Access-Control-Allow-Headers response header is used in response to a
+    /// preflight request which includes the Access-Control-Request-Headers to
+    /// indicate which HTTP headers can be used during the actual request.
+    ///
+    /// Source: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers
+    pub(crate) allow_headers: Option<Vec<String>>,
+    /// The Access-Control-Allow-Methods response header specifies the method or
+    /// methods allowed when accessing the resource in response to a preflight
+    /// request.
+    ///
+    /// Source: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods
+    pub(crate) allow_methods: Option<Vec<String>>,
+    /// The Access-Control-Allow-Origin response header indicates whether the
+    /// response can be shared with requesting code from the given origin.
+    ///
+    /// Source: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
+    pub(crate) allow_origin: Option<String>,
```

Comment:
It could be cool to add the capability to allow several origins. We only can return one Access-Control-Allow-Origin per Response, but could return the origin that we need depending on the origin of the request.

This could be done if we check on runtime the [HTTP_ORIGIN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin) header and if that is one of the allowed origins defined in this configuration, then we can use that as Access-Control-Allow-Origin. Otherwise, return Null in Access-Control-Allow-Origin. 

In order to do that we should extract the hostname from HTTP_ORIGIN. What do you think?

Reply:
I really like the idea, this could help when an instance of this HTTP Server is serving content for a set of particular front-ends.

This PR provides CORS capabilities as defined by the spec, and as you mention above:

```
We only can return one Access-Control-Allow-Origin per Response
```

So I would create a feature request for this and implement it in a near future!

Feel free to open a Feature Request in the issues section, there's a template just for that purposes. I can open it for you if you like anyways! If thats the case, I will also mention that you are the author of this idea of course!

Thanks for our review!